sources/lib/api/gcp/bigquery/_schema.py

```diff
@@ -190,25 +223,48 @@ def from_data(source):
        dataframe.
 
     Args:
-      source: the Pandas Dataframe or list of data from which to infer the schema, or
+      source: the Pandas Dataframe, a dictionary representing a record, a list of heterogeneous
+          data (record) or homogeneous data (list of records) from which to infer the schema, or
           a definition of the schema as a list of dictionaries with 'name' and 'type' entries
           and possibly 'mode' and 'description' entries. Only used if no data argument was provided.
           'mode' can be 'NULLABLE', 'REQUIRED' or 'REPEATED'. For the allowed types, see:
           https://cloud.google.com/bigquery/preparing-data-for-bigquery#datatypes
+
+          Note that there is potential ambiguity when passing a list of lists or a list of
+          dicts between whether that should be treated as a list of records or a single record
+          that is a list. The heuristic used is to check the length of the entries in the
+          list; if they are equal then a list of records is assumed. To avoid this ambuity
+          you can instead use the Schema.from_record method which assumes a single record,
+          in either list of values or dictionary of key-values form.
+
     Returns:
       A Schema for the data.
     """
     if isinstance(source, pandas.DataFrame):
-      return Schema.from_dataframe(source)
+      bq_schema = Schema._from_dataframe(source)
     elif isinstance(source, list):
-      # Inspect the list - if each entry is a dictionary with name and type entries,
-      # then use it directly; else infer from it.
-      if all([isinstance(d, dict) and 'name' in d and 'type' in d for d in source]):
-        return Schema(source)
+      if len(source) == 0:
+        bq_schema = source
+      elif all(isinstance(d, dict) for d in source):
+        if all('name' in d and 'type' in d for d in source):
+          # It looks like a bq_schema; use it as-is.
+          bq_schema = source
+        elif all(len(d) == len(source[0]) for d in source):
+          bq_schema = Schema._from_dict_record(source[0])
+        else:
+          bq_schema = Schema._from_list_record(source)
+      elif len(source) and isinstance(source[0], list) and \
+          all([isinstance(l, list) and len(l) == len(source[0]) for l in source]):
+        # A list of lists all of the same length; treat first entry as a list record.
+        bq_schema = Schema._from_record(source[0])
       else:
-        return Schema.from_list(source)
+        # A heterogeneous list; treat as a record.
+        bq_schema = Schema._from_list_record(source)
+    elif isinstance(source, dict):
```

Comment:
Done this in the three places where we fell back to treating source as a record (not a list, list of heterogeneous dicts/lists).